pubsub/kafka/kafka.go

```diff
@@ -60,35 +61,29 @@ type consumer struct {
 }
 
 func (consumer *consumer) ConsumeClaim(session sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
+	if consumer.callback == nil {
+		return fmt.Errorf("nil consumer callback")
+	}
+
 	bo := backoff.WithContext(consumer.backOff, session.Context())
 	for message := range claim.Messages() {
-		if consumer.callback == nil {
-			return fmt.Errorf("nil consumer callback")
+		msg := pubsub.NewMessage{
+			Topic: message.Topic,
+			Data:  message.Value,
 		}
-
-		var warningLogged bool
-		err := backoff.RetryNotify(func() error {
-			err := consumer.callback(&pubsub.NewMessage{
-				Topic: claim.Topic(),
-				Data:  message.Value,
-			})
+		if err := pubsub.RetryNotifyRecover(func() error {
+			consumer.logger.Debugf("Processing Kafka message: %s/%d/%d [key=%s]", message.Topic, message.Partition, message.Offset, asBase64String(message.Key))
```

Comment:
This means that every message would have the key base64 encoded. Please, remove this part for this log.

Reply:
That's actually not happening here. The reason for the `asBase64String` alias type (that implements [Stringer](https://golang.org/pkg/fmt/#Stringer)) is that it postpones the base64 encoding operation until the log is actually written. Since this is logged at Debug level, base64 is never invoked in the happy path. IMO since the key typically correlates to an entity ID in the user's overall architecture, it provides valuable context in the log.

Reply:
OK. Please, add comment about this to avoid someone reverting this later.

Reply:
Good call. I'll put the comment above `asBase64String`